drivers/i2c/i2c_mcux_flexcomm.c

```diff
@@ -345,18 +423,36 @@ int mcux_flexcomm_target_unregister(const struct device *dev,
 	struct mcux_flexcomm_data *data = dev->data;
 	struct mcux_flexcomm_target_data *target;
 	I2C_Type *base = config->base;
+	uint32_t clock_freq;
+
+	/* Get the clock frequency */
+	if (clock_control_get_rate(config->clock_dev, config->clock_subsys,
+				   &clock_freq)) {
+		return -EINVAL;
+	}
 
 	target = mcux_flexcomm_find_target_by_address(data, target_config->address);
 	if (!target || !target->target_attached) {
 		return -EINVAL;
 	}
 
+	if (mcux_flexcomm_setup_i2c_config_address(data, target, true) < 0) {
+		return -EINVAL;
+	}
+
 	target->target_cfg = NULL;
 	target->target_attached = false;
 
 	data->nr_targets_attached--;
 
-	I2C_SlaveDeinit(base);
+	if (data->nr_targets_attached == 0) {
+		I2C_SlaveDeinit(base);
+	} else {
+		/* still slaves attached, reconfigure the I2C peripheral after address removal */
+		if (mcux_flexcomm_setup_slave_config(dev) < 0) {
+			return -EINVAL;
+		}
+	}
```

Comment:
NIT: you have the special casing in the else clause, it IMHO would be more logical to swap it around, E.G.
```
if (data->nr_targets_attached > 0) {
  	/* still slaves attached, reconfigure the I2C peripheral after address removal */
	if (mcux_flexcomm_setup_slave_config(dev) < 0) {
		return -EINVAL;
	}
} else {
       I2C_SlaveDeinit(base);
}
```